System-wide alert on all monitor pages

DomGarguilo opened the pull request:
This PR adds an alert that appears on the top of every page in the monitor when there is a system-wide message. The only cases where this alert appears right now is when the manager is in a non-normal state. The code is extensible and if other cases are added in the future they can be added to the code. The alert allows for multiple messages at a time and will append them to a list in the alert. Along with the alert banner, a new status notification was added to the navbar which on hover will say that an alert is present and to see the top of the page.

Here is a screen recording where the system was in a normal state, then I switched the manager to `SAFE_MODE` and then to `CLEAN_STOP`. Its not shown in the recording but once the manager shuts down, the alert disapears.
![full](https://github.com/apache/accumulo/assets/47725857/e30819bd-6c15-4447-b3f9-925243ff50bf)

Manno15 reviewed (approved):
The alert looks good. Does seem there is some empty space at the bottom of the alert that could be trimmed so the alert overall is not quite as large. Also, maybe a way to minimize the alert once its shown since it will always be on the navbar as well to illustrate that there is one currently. Thoughts on that?

DomGarguilo commented:
> The alert looks good. Does seem there is some empty space at the bottom of the alert that could be trimmed so the alert overall is not quite as large. 

Good idea. I went ahead and tried that out and think it looks good. Ill upload a screenshot in another comment.

> Also, maybe a way to minimize the alert once its shown since it will always be on the navbar as well to illustrate that there is one currently. Thoughts on that?

I think this is a good idea too but not sure what the best course of action here is. I think it would be better to minimize the alert as opposed to remove it but I am open to other suggestions as well if anyone else has any. Maybe a [boostrap accordion](https://getbootstrap.com/docs/5.3/components/accordion/)-like minimize where the messages can be collapsed into the title "System-wide Notifications"

DomGarguilo commented:
Removed the margin at the bottom of the list of alert messages in 0dc7932. I also tested out adding multiple messages to that list to see how it would look:

![Screenshot from 2023-09-25 09-47-57](https://github.com/apache/accumulo/assets/47725857/b9dbd914-1533-4c77-866b-2c221c4925e7)

Right now the manager states is the only thing that adds to this alert but in the future if we wanted to add others it might look something similar to this.

DomGarguilo commented:
In c7f8c07 I converted the bootstrap `alert` to an `accordion` so that the messages could be collapsed up into the header. Here is how things look now:
![full accordion](https://github.com/apache/accumulo/assets/47725857/6bc34ec4-8255-4f62-84d6-e7cad27804a3)

I can revert back to the alert if someone prefers that better. I think either one works well.

ctubbsii commented:
Wow, you did a good job presenting us with a lot of good options. Here are my thoughts from looking at the example images (didn't look at the code changes at all yet):

1. Bootstrap Icons [has better fonts](https://icons.getbootstrap.com/#icon-font) that we can use for the alert character. Instead of using the SVGs, I'd go ahead and download and include the bootstrap icons font files, and use the CSS classes to style the icon, because that seems easier to work with once the fonts are in place. It's basically adding another (small) dependency to the monitor's external resources, but I think it's worth it, and we can take advantage of some of the other icons later, if we want, once it's added.
2. I don't think we need to explicitly say "System-wide Notifications". I think just having the alert window there is enough. If the icon in the box is the same as in the nav bar, and both are on every page, then the connection is pretty intuitive that it's system-wide. There are [examples of icons in alert boxes in the Bootstrap docs](https://getbootstrap.com/docs/5.0/components/alerts/#icons) that I think look good, and they use the bootstrap icons I mentioned (although, their examples use the SVG instead of the font, which is annoying, because there aren't great examples of using the fonts... even though they should be easier to use).
3. I like the colors and overall appearance of the alert box more than the accordian, but I do really like the ability to minimize or dismiss it. I'm thinking that we should stick with the alert box, but add the [dismiss feature](https://getbootstrap.com/docs/5.0/components/alerts/#dismissing)). Then, we can make it so the alert comes back if you click the icon in the nav bar ("un-dismissing"? :smiley_cat:). I think you'd have to register an action on the dismiss button, so we can track that it has been dismissed in the local storage, so it doesn't come back when you refresh. Making it show by clicking on the icon in the nav bar is easier.

DomGarguilo commented:
> Wow, you did a good job presenting us with a lot of good options. Here are my thoughts from looking at the example images (didn't look at the code changes at all yet):
> 
> 1. Bootstrap Icons [has better fonts](https://icons.getbootstrap.com/#icon-font) that we can use for the alert character. Instead of using the SVGs, I'd go ahead and download and include the bootstrap icons font files, and use the CSS classes to style the icon, because that seems easier to work with once the fonts are in place. It's basically adding another (small) dependency to the monitor's external resources, but I think it's worth it, and we can take advantage of some of the other icons later, if we want, once it's added.
> 2. I don't think we need to explicitly say "System-wide Notifications". I think just having the alert window there is enough. If the icon in the box is the same as in the nav bar, and both are on every page, then the connection is pretty intuitive that it's system-wide. There are [examples of icons in alert boxes in the Bootstrap docs](https://getbootstrap.com/docs/5.0/components/alerts/#icons) that I think look good, and they use the bootstrap icons I mentioned (although, their examples use the SVG instead of the font, which is annoying, because there aren't great examples of using the fonts... even though they should be easier to use).

I went ahead and implemented these first two points in the past could commits. I replaced the existing svgs with their icon equivalents and removed the svgs. I also removed the title on the new alert and just replaced it with the icon. I agree that it does not need the title but was not too sure where to place the icon in relation to the unordered list within the alerts. Here is what things look like now:

 
![Screenshot from 2023-10-02 14-24-53](https://github.com/apache/accumulo/assets/47725857/9c4d2986-7580-431f-a845-66e47d447156)
![Screenshot from 2023-10-02 14-21-44](https://github.com/apache/accumulo/assets/47725857/d1555a08-99c4-4d7e-967e-44e4adb539a3)


> 3. I like the colors and overall appearance of the alert box more than the accordian, but I do really like the ability to minimize or dismiss it. I'm thinking that we should stick with the alert box, but add the [dismiss feature](https://getbootstrap.com/docs/5.0/components/alerts/#dismissing)). Then, we can make it so the alert comes back if you click the icon in the nav bar ("un-dismissing"? 😺). I think you'd have to register an action on the dismiss button, so we can track that it has been dismissed in the local storage, so it doesn't come back when you refresh. Making it show by clicking on the icon in the nav bar is easier.

I'll start working on this part now.

DomGarguilo commented:
8163f59 adds a dismiss button that will dismiss the alert locally (the alert will not be affected on other machines). Clicking the new icon will bring the alert back up. Here is a demo of dismissing and bringing back the alert while switching pages and having auto-refresh enabled:
![dissmiss demo](https://github.com/apache/accumulo/assets/47725857/61a4c40f-e6f1-4d9b-ab9f-cd6f17e72d8a)

ctubbsii commented on server/monitor/src/main/resources/org/apache/accumulo/monitor/templates/default.ftl:
```diff
@@ -45,6 +45,7 @@
       <script src="/resources/external/flot/jquery.flot.time.js"></script>
       <script src="/resources/external/flot/jquery.flot.resize.js"></script>
       <link rel="stylesheet" href="/resources/external/bootstrap/css/bootstrap.css" />
+      <link rel="stylesheet" href="/resources/external/bootstrap/css/bootstrap-icons.css" />
```
This new dependency (bootstrap icons are a separate dependency than the bootstrap bundle) requires an update to the LICENSE files.

From the top of the css file, I see this info to use:
```
/*!
 * Bootstrap Icons v1.11.1 (https://icons.getbootstrap.com/)
 * Copyright 2019-2023 The Bootstrap Authors
 * Licensed under MIT (https://github.com/twbs/icons/blob/main/LICENSE)
 */
```

Related: I noticed flot and jquery have license info in all three of:

* LICENSE
* server/monitor/src/main/appended-resources/META-INF/LICENSE
* assemble/src/main/resources/LICENSE

But bootstrap and datatables are only in the first two, but should probably be in the third as well.

The first one is for the source tarball. The last one is for the binary tarball, and the middle one is for the monitor's jar, which is also contained in the binary tarball, but should still list things at the root of the tarball.

DomGarguilo replied:
Should be fixed in 80a743c

ctubbsii reviewed (changes requested):
Looks good overall. Just a comment on updating the LICENSE files.

I would also like to see if we can tweak the content of the alert window, so the icon is to the left of the list of alerts, instead of above it in a separate section. I didn't investigate too deeply to see how we could do that. The examples on the bootstrap site show it like that for one item, but I think it could work nicely for more than one. I was thinking something like:


```
 ICON  Item 1
 ICON  Item 2
       Item 3
       Item 4
```

In this example, I show "ICON" on two lines, because it'll be a bigger font, so it may span multiple lines. I'm a bit worried about bike-shedding this, so if it's too much trouble, then don't worry about the icon placement that much.

DomGarguilo commented:
> Looks good overall. Just a comment on updating the LICENSE files.
> 
> I would also like to see if we can tweak the content of the alert window, so the icon is to the left of the list of alerts, instead of above it in a separate section. I didn't investigate too deeply to see how we could do that. The examples on the bootstrap site show it like that for one item, but I think it could work nicely for more than one. I was thinking something like:
> 
> ```
>  ICON  Item 1
>  ICON  Item 2
>        Item 3
>        Item 4
> ```
> 
> In this example, I show "ICON" on two lines, because it'll be a bigger font, so it may span multiple lines. I'm a bit worried about bike-shedding this, so if it's too much trouble, then don't worry about the icon placement that much.

Tried this in 46e3e9c. I think it looks good:
![Screenshot from 2023-10-04 14-53-40](https://github.com/apache/accumulo/assets/47725857/c9497720-2b1e-4e65-b38f-8a124f87d16e)
And with multiple items in the list:
![Screenshot from 2023-10-04 14-53-29](https://github.com/apache/accumulo/assets/47725857/efced79d-ddec-459c-98a5-f012944e15d5)

ctubbsii commented on server/monitor/src/main/appended-resources/META-INF/LICENSE:
```diff
@@ -75,14 +75,23 @@ Files:
      * http://benalman.com/about/license/
 
-## Bootstrap v5.3.1 (https://getbootstrap.com/)
+## Bootstrap v5.3.2 (https://getbootstrap.com/)
```
I must be going crazy, because I thought I had checked these already. But nice catch.

DomGarguilo replied:
I think we had discussed who would fix this on the bootstrap ticket but neither of us ever got around to saying that we would do it.

ctubbsii replied:
I definitely remember checking. I definitely re-downloaded the 5.3.2 versions, to make sure they matched bit-for-bit. I thought I checked the version in the LICENSE files at the same time, but maybe I saw 5.3.1 and my brain just replaced the 1 with a 2 in my head. :shrug:

Manno15 reviewed (approved):
I like the recent changes.
